src/daemon/daemon_rpc.cpp

```diff
@@ -92,143 +90,215 @@ grpc::Status emit_signal_and_wait_for_result(OperationSignal operation_signal)
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_stop, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::suspend(grpc::ServerContext* context, const SuspendRequest* request,
                                     grpc::ServerWriter<SuspendReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_suspend, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_suspend, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::restart(grpc::ServerContext* context, const RestartRequest* request,
                                     grpc::ServerWriter<RestartReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_restart, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_restart, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::delet(grpc::ServerContext* context, const DeleteRequest* request,
                                   grpc::ServerWriter<DeleteReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_delete, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_delete, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::umount(grpc::ServerContext* context, const UmountRequest* request,
                                    grpc::ServerWriter<UmountReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::version(grpc::ServerContext* context, const VersionRequest* request,
                                     grpc::ServerWriter<VersionReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::ping(grpc::ServerContext* context, const PingRequest* request, PingReply* response)
 {
+    auto client_cert = client_cert_from(context);
+
+    if (!client_cert.empty() && !client_cert_store->verify_cert(client_cert))
+    {
+        return grpc::Status{grpc::StatusCode::UNAUTHENTICATED, ""};
+    }
+
     return grpc::Status::OK;
```

Comment:
Would somehow feel safer if this logic was reversed (only OK _if_ all good).

Reply:
Ok, sure.

Reply:
Fixed